### ISSUE TITLE
Add a Data wrapper for edge_index, x, y and edge_attr

### ISSUE THREAD

**Issue by TristanBilot** [TristanBilot/mlx-graphs#6]
It should be good practice to use a Data-like data structure like in PyG, to compute some operations using this structure instead of passing all args separately.

I heard you wanted to start working on that @kheims?

[Here](https://github.com/pyg-team/pytorch_geometric/blob/5c561949d67123cf6ea892812e2357ce9bd9a819/torch_geometric/data/data.py) is how it was done in PyG few years ago

**Comment by TristanBilot** [comment]
Before starting the implem, maybe we should think about the best architecture for this `Data` object?
@gboduljak, @francescofarina

**Comment by Kheims** [comment]
Yes I can start working on it as soon as we come up with the best architecture for it.
For now i'll begin making a scaffolding for formalities ( class's attributes as they will probably have the same as pyg's one as well as methods like number_of_nodes and number_of_edges )

**Comment by gboduljak** [comment]
> Before starting the implem, maybe we should think about the best architecture for this `Data` object? @gboduljak, @francescofarina

I think we should match PyG ``Data`` functionality. We might call this ``GraphData`` instead, but this is a minor. Ideally, this should be very small and generic. I think it is useful to think about how to provide generic ``collate_fn`` for graph batching while we design this object. The object should be designed to implement generic ``collate_fn`` in the easiest possible way.

**Comment by francescofarina** [comment]
Let's perhaps try to adopt a better naming and formalism than PyG. For example I find the names of the attributes of `Data` in PyG misleading and not self-explanatory; e.g., I'd rather have `x` -> `node_features`, `edge_attr` -> `edge_features`

**Comment by TristanBilot** [comment]
I came up with an idea regarding the `Data` object. 
`Data` in PyG can be confusing as we don't really know what's inside. 
This data model could be more intuitive: `Dataset` -> `Graphs` -> `Graph`

* **Graph**: similar to `Data` in PyG, contains all the features and functions for a single graph.
* **Graphs**: similar to `Batch` in PyG, contains a list of `Graph` and functions for batching, etc. (maybe change `Graphs` by `Batch` if less intuitive).
* **Dataset**: contains a list of `Graphs` or a list of `Graphs`, and adapters to convert `Dataset` from PyG within a `Dataset` for mlx. At the beginning, we could use only load data using PyG as backend, and remove this dependency in the future?

**Comment by gboduljak** [comment]
> I came up with an idea regarding the `Data` object. `Data` in PyG can be confusing as we don't really know what's inside. This data model could be more intuitive: `Dataset` -> `Graphs` -> `Graph`
> 
> * **Graph**: similar to `Data` in PyG, contains all the features and functions for a single graph.
> * **Graphs**: similar to `Batch` in PyG, contains a list of `Graph` and functions for batching, etc. (maybe change `Graphs` by `Batch` if less intuitive).
> * **Dataset**: contains a list of `Graphs` or a list of `Graph`, and adapters to convert `Dataset` from PyG into a `Dataset` for mlx. At the beginning, we could use only load data using PyG as backend, and remove this dependency in the future?
> 
> ```
> g1 = Graph(node_feats=..., edge_index=...,)
> g2 = Graph(node_feats=..., edge_index=...,)
> 
> graphs = Graphs([g1, g2], batch_size=2)
> ...
> h = gnn(graphs)
> ```

So ``Graphs`` does the graph batching?

**Comment by gboduljak** [comment]
> I came up with an idea regarding the `Data` object. `Data` in PyG can be confusing as we don't really know what's inside. This data model could be more intuitive: `Dataset` -> `Graphs` -> `Graph`
> 
> * **Graph**: similar to `Data` in PyG, contains all the features and functions for a single graph.
> * **Graphs**: similar to `Batch` in PyG, contains a list of `Graph` and functions for batching, etc. (maybe change `Graphs` by `Batch` if less intuitive).
> * **Dataset**: contains a list of `Graphs` or a list of `Graph`, and adapters to convert `Dataset` from PyG into a `Dataset` for mlx. At the beginning, we could use only load data using PyG as backend, and remove this dependency in the future?
> 
> ```
> g1 = Graph(node_feats=..., edge_index=...,)
> g2 = Graph(node_feats=..., edge_index=...,)
> 
> graphs = Graphs([g1, g2], batch_size=2)
> ...
> h = gnn(graphs)
> ```

I would start without PyG loaders.

**Comment by TristanBilot** [comment]
> So `Graphs` does the graph batching?

yup

**Comment by TristanBilot** [comment]
> I would start without PyG loaders.
So we let users load datasets by themselves?

**Comment by gboduljak** [comment]
This looks good to me.

**Comment by gboduljak** [comment]
> > I would start without PyG loaders.
> 
> So we let users load datasets by themselves?

Yes. We should provide utilities for converting between adjacency matrix and edge index.

**Comment by TristanBilot** [comment]
For the batching part, I saw that PyG theoretically stacks multiple adj matrices on the diagonal and computes message passing using scatter on this huge matrix ([source](https://pytorch-geometric.readthedocs.io/en/latest/get_started/introduction.html#mini-batches)), but idk how it deals with it when using edge_index and not ajd matrices, we have to check this.

**Comment by TristanBilot** [comment]
I just thought of this: we have to let the user the possibility to add custom attributes to the `GraphData`, not only node_features, edge_features, graph_features. And those custom attributes have to be handled properly by the batching.

I mean that it would be nice to find a general batching function that can concatenate any custom features added by the user with different shapes.

Ex:

```python
g1 = mxg.graph(edge_index, node_features, custom_src_edge_features=custom_src_edge_features1)
g2 = mxg.graph(edge_index, node_features, custom_src_edge_features=custom_src_edge_features2)

batch = mxg.batch([g1, g2])
batch.custom_src_edge_features # -> should be concatenated along the 0-axis, as for node features
```
For each graph in the list of graphs, we have to check all graphs have the same custom arguments.

**Comment by francescofarina** [comment]
Yes, for sure. I assume the final collate function that'll be implemented in https://github.com/TristanBilot/mlx-graphs/pull/24 will take care of collating any attributes of `GraphData`?

**Comment by TristanBilot** [comment]
Yes, I think it makes sense.

**Comment by TristanBilot** [comment]
But how to handle specific cases where a user want to concat along a different axis? Like the 1-axis for edge_index, or using a stack with a new 0-axis, etc.

**Comment by francescofarina** [comment]
I think we can be prescriptive about it (e.g. we decide how concatenation is handled by default or we implement it only for the default set of attributes) and then we allow the user to overwrite the collate function to handle their specific needs.

**Comment by francescofarina** [comment]
Data and batching implemented in #39. Closing this and opening one for datasets/dataloaders.
